File: swift/Sources/ProjectNameSDK/core/BluefinPayconexSDK.swift
// BluefinPayconex SDK client.

import Foundation

public final class BluefinPayconexSDK {
  public var mode = "live"
  private var options: VMap = VMap()
  private let utility: Utility
  public var features: [BaseFeature] = []
  private var rootctx: Context!

  public init(_ optionsIn: VMap? = nil) {
    utility = Utility()

    let config = SdkConfig.makeConfig()

    var ctxmap: [String: Any?] = [
      "client": self,
      "utility": utility,
      "config": config,
      "shared": VMap(),
    ]
    if let o = optionsIn { ctxmap["options"] = o }

    rootctx = utility.makeContext(ctxmap, nil)

    options = utility.makeOptions(rootctx)

    if gpath(options, "feature", "test", "active") == .bool(true) {
      mode = "test"
    }

    rootctx.options = options

    // Add features in the resolved order (makeOptions puts an explicit list
    // order first, else defaults to test-first). Ordering matters: the `test`
    // feature installs the base mock transport and the transport features
    // (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    // must be added before them to sit at the base of the chain.
    let featureOpts = gp(options, "feature").asMap ?? VMap()
    if let featureOrder = gpath(options, "__derived__", "featureorder").asList {
      for fnameVal in featureOrder.items {
        let fname = fnameVal.asString ?? ""
        if fname != "", let fopts = gp(featureOpts, fname).asMap,
          fopts.entries["active"]?.asBool == true {
          utility.featureAdd(rootctx, SdkConfig.makeFeature(fname))
        }
      }
    }

    // Add extension features.
    if let extList = gp(options, "extend").asList {
      for f in extList.items {
        if let feat = f.asNative as? BaseFeature {
          utility.featureAdd(rootctx, feat)
        }
      }
    }

    // Initialize features.
    for f in features {
      utility.featureInit(rootctx, f)
    }

    utility.featureHook(rootctx, "PostConstruct")
  }

  public func optionsMap() -> VMap {
    return clone(.map(options)).asMap ?? VMap()
  }

  public func getUtility() -> Utility {
    return Utility.copy(utility)
  }

  public func getRootCtx() -> Context {
    return rootctx
  }

  public func prepare(_ fetchargsIn: VMap?) throws -> VMap {
    let utility = self.utility

    let fetchargs = fetchargsIn ?? VMap()

    let ctrl = gp(fetchargs, "ctrl").asMap ?? VMap()

    let ctx = utility.makeContext(["opname": "prepare", "ctrl": ctrl], rootctx)

    let options = self.options

    let path = gp(fetchargs, "path").asString ?? ""
    var method = gp(fetchargs, "method").asString ?? ""
    if method == "" { method = "GET" }

    let pathParams = gp(fetchargs, "params").asMap ?? VMap()
    let query = gp(fetchargs, "query").asMap ?? VMap()

    let headers = utility.prepareHeaders(ctx)

    let basev = gp(options, "base").asString ?? ""
    let prefix = gp(options, "prefix").asString ?? ""
    let suffix = gp(options, "suffix").asString ?? ""

    let specmap = VMap()
    specmap.entries["base"] = .string(basev)
    specmap.entries["prefix"] = .string(prefix)
    specmap.entries["suffix"] = .string(suffix)
    specmap.entries["path"] = .string(path)
    specmap.entries["method"] = .string(method)
    specmap.entries["params"] = .map(pathParams)
    specmap.entries["query"] = .map(query)
    specmap.entries["headers"] = .map(headers)
    specmap.entries["body"] = gp(fetchargs, "body")
    specmap.entries["step"] = .string("start")
    ctx.spec = Spec(specmap)

    // Merge user-provided headers.
    if let uhm = gp(fetchargs, "headers").asMap {
      for (k, v) in uhm.entries {
        ctx.spec!.headers.entries[k] = v
      }
    }

    _ = try utility.prepareAuth(ctx)

    return try utility.makeFetchDef(ctx)
  }

  public func direct(_ fetchargsIn: VMap?) -> VMap {
    let utility = self.utility

    let fetchdef: VMap
    do {
      fetchdef = try prepare(fetchargsIn)
    } catch {
      let r = VMap()
      r.entries["ok"] = .bool(false)
      r.entries["err"] = .nat(error)
      return r
    }

    let fetchargs = fetchargsIn ?? VMap()
    let ctrl = gp(fetchargs, "ctrl").asMap ?? VMap()

    let ctx = utility.makeContext(["opname": "direct", "ctrl": ctrl], rootctx)

    let url = gp(fetchdef, "url").asString ?? ""

    let fetched: Value
    do {
      fetched = try utility.fetcher(ctx, url, fetchdef)
    } catch {
      let r = VMap()
      r.entries["ok"] = .bool(false)
      r.entries["err"] = .nat(error)
      return r
    }

    if isNil(fetched) {
      let r = VMap()
      r.entries["ok"] = .bool(false)
      r.entries["err"] = .nat(ctx.makeError("direct_no_response", "response: undefined"))
      return r
    }

    if let fm = fetched.asMap {
      let status = toInt(gp(fm, "status"))
      let headers = gp(fm, "headers")

      // No-body responses (204, 304) and explicit zero content-length must
      // skip JSON parsing.
      var contentLength = ""
      if let hm = headers.asMap, let cl = hm.entries["content-length"], !isNil(cl) {
        contentLength = stringify(cl)
      }
      let noBody = status == 204 || status == 304 || contentLength == "0"

      var jsonData: Value = .noval
      if !noBody, let jf = gp(fm, "json").asNative as? NativeCall0 {
        jsonData = jf()
      }

      let r = VMap()
      r.entries["ok"] = .bool(status >= 200 && status < 300)
      r.entries["status"] = .int(Int64(status))
      r.entries["headers"] = headers
      r.entries["data"] = jsonData
      return r
    }

    let r = VMap()
    r.entries["ok"] = .bool(false)
    r.entries["err"] = .nat(ctx.makeError("direct_invalid", "invalid response type"))
    return r
  }


  // AccountUpdater returns a AccountUpdater entity bound to this client.
  // Idiomatic usage: try client.AccountUpdater().list(nil) or
  // try client.AccountUpdater().load(vm(("id", .string("..."))), nil).
  public func AccountUpdater(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return AccountUpdaterEntity(self, entopts)
  }

  // AccountUpdaterSchedule returns a AccountUpdaterSchedule entity bound to this client.
  // Idiomatic usage: try client.AccountUpdaterSchedule().list(nil) or
  // try client.AccountUpdaterSchedule().load(vm(("id", .string("..."))), nil).
  public func AccountUpdaterSchedule(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return AccountUpdaterScheduleEntity(self, entopts)
  }

  // AccountUpdaterScheduleWithResult returns a AccountUpdaterScheduleWithResult entity bound to this client.
  // Idiomatic usage: try client.AccountUpdaterScheduleWithResult().list(nil) or
  // try client.AccountUpdaterScheduleWithResult().load(vm(("id", .string("..."))), nil).
  public func AccountUpdaterScheduleWithResult(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return AccountUpdaterScheduleWithResultEntity(self, entopts)
  }

  // AccountUpdaterSubscriptionWithResult returns a AccountUpdaterSubscriptionWithResult entity bound to this client.
  // Idiomatic usage: try client.AccountUpdaterSubscriptionWithResult().list(nil) or
  // try client.AccountUpdaterSubscriptionWithResult().load(vm(("id", .string("..."))), nil).
  public func AccountUpdaterSubscriptionWithResult(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return AccountUpdaterSubscriptionWithResultEntity(self, entopts)
  }

  // AccountUpdaterUpdate returns a AccountUpdaterUpdate entity bound to this client.
  // Idiomatic usage: try client.AccountUpdaterUpdate().list(nil) or
  // try client.AccountUpdaterUpdate().load(vm(("id", .string("..."))), nil).
  public func AccountUpdaterUpdate(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return AccountUpdaterUpdateEntity(self, entopts)
  }

  // ApiKey returns a ApiKey entity bound to this client.
  // Idiomatic usage: try client.ApiKey().list(nil) or
  // try client.ApiKey().load(vm(("id", .string("..."))), nil).
  public func ApiKey(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return ApiKeyEntity(self, entopts)
  }

  // ApplePayMerchantDetail returns a ApplePayMerchantDetail entity bound to this client.
  // Idiomatic usage: try client.ApplePayMerchantDetail().list(nil) or
  // try client.ApplePayMerchantDetail().load(vm(("id", .string("..."))), nil).
  public func ApplePayMerchantDetail(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return ApplePayMerchantDetailEntity(self, entopts)
  }

  // ApplePaySession returns a ApplePaySession entity bound to this client.
  // Idiomatic usage: try client.ApplePaySession().list(nil) or
  // try client.ApplePaySession().load(vm(("id", .string("..."))), nil).
  public func ApplePaySession(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return ApplePaySessionEntity(self, entopts)
  }

  // DynamicDescriptor returns a DynamicDescriptor entity bound to this client.
  // Idiomatic usage: try client.DynamicDescriptor().list(nil) or
  // try client.DynamicDescriptor().load(vm(("id", .string("..."))), nil).
  public func DynamicDescriptor(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return DynamicDescriptorEntity(self, entopts)
  }

  // IFrameCreateInstance returns a IFrameCreateInstance entity bound to this client.
  // Idiomatic usage: try client.IFrameCreateInstance().list(nil) or
  // try client.IFrameCreateInstance().load(vm(("id", .string("..."))), nil).
  public func IFrameCreateInstance(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return IFrameCreateInstanceEntity(self, entopts)
  }

  // IFrameInstance returns a IFrameInstance entity bound to this client.
  // Idiomatic usage: try client.IFrameInstance().list(nil) or
  // try client.IFrameInstance().load(vm(("id", .string("..."))), nil).
  public func IFrameInstance(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return IFrameInstanceEntity(self, entopts)
  }

  // Iframe returns a Iframe entity bound to this client.
  // Idiomatic usage: try client.Iframe().list(nil) or
  // try client.Iframe().load(vm(("id", .string("..."))), nil).
  public func Iframe(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return IframeEntity(self, entopts)
  }

  // Init returns a Init entity bound to this client.
  // Idiomatic usage: try client.Init().list(nil) or
  // try client.Init().load(vm(("id", .string("..."))), nil).
  public func Init(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return InitEntity(self, entopts)
  }

  // ListApiKeyScopesEntry returns a ListApiKeyScopesEntry entity bound to this client.
  // Idiomatic usage: try client.ListApiKeyScopesEntry().list(nil) or
  // try client.ListApiKeyScopesEntry().load(vm(("id", .string("..."))), nil).
  public func ListApiKeyScopesEntry(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return ListApiKeyScopesEntryEntity(self, entopts)
  }

  // PaymentIframe returns a PaymentIframe entity bound to this client.
  // Idiomatic usage: try client.PaymentIframe().list(nil) or
  // try client.PaymentIframe().load(vm(("id", .string("..."))), nil).
  public func PaymentIframe(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return PaymentIframeEntity(self, entopts)
  }

  // ThreeDSecureAuth returns a ThreeDSecureAuth entity bound to this client.
  // Idiomatic usage: try client.ThreeDSecureAuth().list(nil) or
  // try client.ThreeDSecureAuth().load(vm(("id", .string("..."))), nil).
  public func ThreeDSecureAuth(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return ThreeDSecureAuthEntity(self, entopts)
  }

  // ThreeDSecureBrowserInit returns a ThreeDSecureBrowserInit entity bound to this client.
  // Idiomatic usage: try client.ThreeDSecureBrowserInit().list(nil) or
  // try client.ThreeDSecureBrowserInit().load(vm(("id", .string("..."))), nil).
  public func ThreeDSecureBrowserInit(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return ThreeDSecureBrowserInitEntity(self, entopts)
  }

  // ThreeDSecureStatus returns a ThreeDSecureStatus entity bound to this client.
  // Idiomatic usage: try client.ThreeDSecureStatus().list(nil) or
  // try client.ThreeDSecureStatus().load(vm(("id", .string("..."))), nil).
  public func ThreeDSecureStatus(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return ThreeDSecureStatusEntity(self, entopts)
  }

  // TransactionDetail returns a TransactionDetail entity bound to this client.
  // Idiomatic usage: try client.TransactionDetail().list(nil) or
  // try client.TransactionDetail().load(vm(("id", .string("..."))), nil).
  public func TransactionDetail(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return TransactionDetailEntity(self, entopts)
  }

  // Webhook returns a Webhook entity bound to this client.
  // Idiomatic usage: try client.Webhook().list(nil) or
  // try client.Webhook().load(vm(("id", .string("..."))), nil).
  public func Webhook(_ entopts: VMap? = nil) -> BluefinPayconexEntityBase {
    return WebhookEntity(self, entopts)
  }


  public static func testSDK(_ testoptsIn: VMap?, _ sdkoptsIn: VMap?) -> BluefinPayconexSDK {
    let sdkopts = clone(.map(sdkoptsIn ?? VMap())).asMap ?? VMap()

    let testopts = clone(.map(testoptsIn ?? VMap())).asMap ?? VMap()
    testopts.entries["active"] = .bool(true)

    _ = setpath(.map(sdkopts), jtp("feature", "test"), .map(testopts))

    let sdk = BluefinPayconexSDK(sdkopts)
    sdk.mode = "test"
    return sdk
  }
}

File: lua/bluefin-payconex_sdk.lua
-- BluefinPayconex SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Typed-model annotations (LuaLS ---@class); empty at runtime.
require("bluefin-payconex_types")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local BluefinPayconexSDK = {}
BluefinPayconexSDK.__index = BluefinPayconexSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

BluefinPayconexSDK._make_feature = _make_feature


function BluefinPayconexSDK.new(options)
  local self = setmetatable({}, BluefinPayconexSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

    -- feature: test


  return self
end


function BluefinPayconexSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function BluefinPayconexSDK:get_utility()
  return Utility.copy(self._utility)
end


function BluefinPayconexSDK:get_root_ctx()
  return self._rootctx
end


function BluefinPayconexSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function BluefinPayconexSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:AccountUpdater():list() / client:AccountUpdater():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:AccountUpdater(data)
  local EntityMod = require("entity.account_updater_entity")
  if data == nil then
    if self._account_updater == nil then
      self._account_updater = EntityMod.new(self, nil)
    end
    return self._account_updater
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:AccountUpdaterSchedule():list() / client:AccountUpdaterSchedule():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:AccountUpdaterSchedule(data)
  local EntityMod = require("entity.account_updater_schedule_entity")
  if data == nil then
    if self._account_updater_schedule == nil then
      self._account_updater_schedule = EntityMod.new(self, nil)
    end
    return self._account_updater_schedule
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:AccountUpdaterScheduleWithResult():list() / client:AccountUpdaterScheduleWithResult():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:AccountUpdaterScheduleWithResult(data)
  local EntityMod = require("entity.account_updater_schedule_with_result_entity")
  if data == nil then
    if self._account_updater_schedule_with_result == nil then
      self._account_updater_schedule_with_result = EntityMod.new(self, nil)
    end
    return self._account_updater_schedule_with_result
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:AccountUpdaterSubscriptionWithResult():list() / client:AccountUpdaterSubscriptionWithResult():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:AccountUpdaterSubscriptionWithResult(data)
  local EntityMod = require("entity.account_updater_subscription_with_result_entity")
  if data == nil then
    if self._account_updater_subscription_with_result == nil then
      self._account_updater_subscription_with_result = EntityMod.new(self, nil)
    end
    return self._account_updater_subscription_with_result
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:AccountUpdaterUpdate():list() / client:AccountUpdaterUpdate():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:AccountUpdaterUpdate(data)
  local EntityMod = require("entity.account_updater_update_entity")
  if data == nil then
    if self._account_updater_update == nil then
      self._account_updater_update = EntityMod.new(self, nil)
    end
    return self._account_updater_update
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ApiKey():list() / client:ApiKey():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:ApiKey(data)
  local EntityMod = require("entity.api_key_entity")
  if data == nil then
    if self._api_key == nil then
      self._api_key = EntityMod.new(self, nil)
    end
    return self._api_key
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ApplePayMerchantDetail():list() / client:ApplePayMerchantDetail():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:ApplePayMerchantDetail(data)
  local EntityMod = require("entity.apple_pay_merchant_detail_entity")
  if data == nil then
    if self._apple_pay_merchant_detail == nil then
      self._apple_pay_merchant_detail = EntityMod.new(self, nil)
    end
    return self._apple_pay_merchant_detail
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ApplePaySession():list() / client:ApplePaySession():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:ApplePaySession(data)
  local EntityMod = require("entity.apple_pay_session_entity")
  if data == nil then
    if self._apple_pay_session == nil then
      self._apple_pay_session = EntityMod.new(self, nil)
    end
    return self._apple_pay_session
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:DynamicDescriptor():list() / client:DynamicDescriptor():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:DynamicDescriptor(data)
  local EntityMod = require("entity.dynamic_descriptor_entity")
  if data == nil then
    if self._dynamic_descriptor == nil then
      self._dynamic_descriptor = EntityMod.new(self, nil)
    end
    return self._dynamic_descriptor
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:IFrameCreateInstance():list() / client:IFrameCreateInstance():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:IFrameCreateInstance(data)
  local EntityMod = require("entity.i_frame_create_instance_entity")
  if data == nil then
    if self._i_frame_create_instance == nil then
      self._i_frame_create_instance = EntityMod.new(self, nil)
    end
    return self._i_frame_create_instance
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:IFrameInstance():list() / client:IFrameInstance():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:IFrameInstance(data)
  local EntityMod = require("entity.i_frame_instance_entity")
  if data == nil then
    if self._i_frame_instance == nil then
      self._i_frame_instance = EntityMod.new(self, nil)
    end
    return self._i_frame_instance
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Iframe():list() / client:Iframe():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:Iframe(data)
  local EntityMod = require("entity.iframe_entity")
  if data == nil then
    if self._iframe == nil then
      self._iframe = EntityMod.new(self, nil)
    end
    return self._iframe
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Init():list() / client:Init():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:Init(data)
  local EntityMod = require("entity.init_entity")
  if data == nil then
    if self._init == nil then
      self._init = EntityMod.new(self, nil)
    end
    return self._init
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ListApiKeyScopesEntry():list() / client:ListApiKeyScopesEntry():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:ListApiKeyScopesEntry(data)
  local EntityMod = require("entity.list_api_key_scopes_entry_entity")
  if data == nil then
    if self._list_api_key_scopes_entry == nil then
      self._list_api_key_scopes_entry = EntityMod.new(self, nil)
    end
    return self._list_api_key_scopes_entry
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PaymentIframe():list() / client:PaymentIframe():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:PaymentIframe(data)
  local EntityMod = require("entity.payment_iframe_entity")
  if data == nil then
    if self._payment_iframe == nil then
      self._payment_iframe = EntityMod.new(self, nil)
    end
    return self._payment_iframe
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ThreeDSecureAuth():list() / client:ThreeDSecureAuth():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:ThreeDSecureAuth(data)
  local EntityMod = require("entity.three_d_secure_auth_entity")
  if data == nil then
    if self._three_d_secure_auth == nil then
      self._three_d_secure_auth = EntityMod.new(self, nil)
    end
    return self._three_d_secure_auth
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ThreeDSecureBrowserInit():list() / client:ThreeDSecureBrowserInit():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:ThreeDSecureBrowserInit(data)
  local EntityMod = require("entity.three_d_secure_browser_init_entity")
  if data == nil then
    if self._three_d_secure_browser_init == nil then
      self._three_d_secure_browser_init = EntityMod.new(self, nil)
    end
    return self._three_d_secure_browser_init
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ThreeDSecureStatus():list() / client:ThreeDSecureStatus():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:ThreeDSecureStatus(data)
  local EntityMod = require("entity.three_d_secure_status_entity")
  if data == nil then
    if self._three_d_secure_status == nil then
      self._three_d_secure_status = EntityMod.new(self, nil)
    end
    return self._three_d_secure_status
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:TransactionDetail():list() / client:TransactionDetail():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:TransactionDetail(data)
  local EntityMod = require("entity.transaction_detail_entity")
  if data == nil then
    if self._transaction_detail == nil then
      self._transaction_detail = EntityMod.new(self, nil)
    end
    return self._transaction_detail
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Webhook():list() / client:Webhook():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function BluefinPayconexSDK:Webhook(data)
  local EntityMod = require("entity.webhook_entity")
  if data == nil then
    if self._webhook == nil then
      self._webhook = EntityMod.new(self, nil)
    end
    return self._webhook
  end
  return EntityMod.new(self, data)
end




function BluefinPayconexSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = BluefinPayconexSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return BluefinPayconexSDK
